qiita_db/archive.py

```diff
@@ -0,0 +1,178 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
+#
+# Distributed under the terms of the BSD 3-clause License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+from __future__ import division
+
+import qiita_db as qdb
+
+
+class Archive(qdb.base.QiitaObject):
+    r"""Extra information for any features stored in a BIOM Artifact
+
+    Methods
+    -------
+    insert_from_biom
+
+    See Also
+    --------
+    qiita_db.QiitaObject
+    """
+
+    @classmethod
+    def _inserting_main_steps(cls, TRN, ms, features):
+        sql = """INSERT INTO qiita.archive_merging_scheme
+                    (archive_merging_scheme)
+                 SELECT %s WHERE NOT EXISTS (
+                    SELECT 1 FROM qiita.archive_merging_scheme
+                    WHERE archive_merging_scheme = %s)"""
+        TRN.add(sql, [ms, ms])
+        sql = """SELECT archive_merging_scheme_id
+                 FROM qiita.archive_merging_scheme
+                 WHERE archive_merging_scheme = %s"""
+        TRN.add(sql, [ms])
+        amsi = TRN.execute_fetchlast()
+
+        vals = [[amsi, _id, val] for _id, val in features.items()]
+        TRN.add("SELECT archive_upsert(%s, %s, %s)", vals, many=True)
+        TRN.execute()
+
+    @classmethod
+    def insert_from_artifact(cls, artifact, features):
+        r"""Inserts new features to the database based on a given artifact
+
+        Parameters
+        ----------
+        artifact : qiita_db.artifact.Artifact
+            The artifact from which the features were generated
+        features : dict {str: str}
+            A dictionary of the features and the values to be stored
+
+        Raises
+        ------
+            ValueError
+                If the Artifact type is not BIOM
+                If the artifact doesn't have a biom filepath
+        """
+        with qdb.sql_connection.TRN:
```

Comment:
seems like most of the logic contained can be done outside of the context?

Reply:
possible but I think it's better to leave it like this as some of these steps are actually accessing the DB so this assures that everything it's in the same transaction.

Reply:
oh. implicit operations remote resource operations? :/ okay.